### PR TITLE
tests: Add the S4U proof of concept as a verify test

### PR DESCRIPTION
**Pull request by Gundersanne**
- [x] blocked because debian-stable needs python-gssapi

### CODE REVIEW

**Comment by martinpitt** [comment]
Some failures are just s/python/python3/ (I suppose), but on Debian/Ubuntu ssh doesn't seem to accept the s4u ticket?

File: test/verify/check-s4u-ssh [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        # set hostname for readable logs
+        client_machine.execute("hostnamectl set-hostname sshclient.cockpit.lan")
+        sshd_machine.execute("hostnamectl set-hostname sshserver.cockpit.lan")
+        self.login_and_go("/system")
+
+        # join both machines
+        wait(lambda: client_machine.execute("nslookup -type=SRV _ldap._tcp.cockpit.lan"))
+        wait(lambda: sshd_machine.execute("nslookup -type=SRV _ldap._tcp.cockpit.lan"))
+        wait(lambda: client_machine.execute("echo foobarfoo | realm join -U admin cockpit.lan"))
+        wait(lambda: sshd_machine.execute("echo foobarfoo | realm join -U admin cockpit.lan"))
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:51]: Please don't use wait() for these -- it really shouldn't run `realm join` more than once, and it's not expected to fail, just take a while.

File: test/verify/check-s4u-ssh [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        # add user to impersonate
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa user-add --first=user --last=user user'")
+        # ipa_machine.execute("echo foobarfoo | ipa user-mod user --password")
+        # add service which will be allowed to delegate creds
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa service-add cockpitclient/sshclient.cockpit.lan@COCKPIT.LAN --ok-as-delegate=true --ok-to-auth-as-delegate=true'")
+        # Allow restrieval of service/host keytabs by admin
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa service-allow-retrieve-keytab --user=admin cockpitclient/sshclient.cockpit.lan'")
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa host-allow-retrieve-keytab --user=admin sshserver.cockpit.lan'")
+        # set up delegation rule
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa servicedelegationtarget-add cockpit-target'")
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:64]: Could all these commands be put into a `script = ` constant and then run in one go? That will be a lot less overhead and logging noise.

File: test/verify/check-s4u-ssh [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+        # replace with
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:66]: ... what?

File: test/verify/check-s4u-ssh [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa servicedelegationtarget-add-member cockpit-target --principals=\"host/sshserver.cockpit.lan@COCKPIT.LAN\"'")
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa servicedelegationrule-add cockpit-delegation'")
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa servicedelegationrule-add-member cockpit-delegation --principals=\"cockpitclient/sshclient.cockpit.lan@COCKPIT.LAN\"'")
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa servicedelegationrule-add-target cockpit-delegation --servicedelegationtargets=\"cockpit-target\"'")
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:70]: Our services image has the [latest freeipa-server docker image](https://hub.docker.com/r/freeipa/freeipa-server/tags) from 12 days ago. Does the absence of the ldap patching hack mean that this has the expected delegation fix now?

File: test/verify/check-s4u-ssh [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        wait(lambda: client_machine.execute("nslookup -type=SRV _ldap._tcp.cockpit.lan"))
+        wait(lambda: sshd_machine.execute("nslookup -type=SRV _ldap._tcp.cockpit.lan"))
+        wait(lambda: client_machine.execute("echo foobarfoo | realm join -U admin cockpit.lan"))
+        wait(lambda: sshd_machine.execute("echo foobarfoo | realm join -U admin cockpit.lan"))
+
+        # configure ipa
+        ipa_machine.execute("docker exec freeipa bash -c 'echo foobarfoo | kinit admin@COCKPIT.LAN'")
+        # add user to impersonate
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa user-add --first=user --last=user user'")
+        # ipa_machine.execute("echo foobarfoo | ipa user-mod user --password")
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:57]: That's for being able to log in to debug interactively? I suppose that can stay.

File: test/verify/check-s4u-ssh [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        # add service which will be allowed to delegate creds
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa service-add cockpitclient/sshclient.cockpit.lan@COCKPIT.LAN --ok-as-delegate=true --ok-to-auth-as-delegate=true'")
+        # Allow restrieval of service/host keytabs by admin
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:60]: re"t"rieval. That looks like some weird kind of "unbreak my setup" step -- these commands are already run as user "admin", thus it has the power to grant that privilege to itself. I suppose the "by admin" is not the important part here, but the "by the involved services"?

What does the  `--user admin` option actually mean here? That a service like `cockpitclient/sshclient.cockpit.lan` can retrieve the keytab *of* the "admin" user? I. e. that's the moral equivalent of setuid root?

File: test/verify/check-s4u-ssh [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa service-allow-retrieve-keytab --user=admin cockpitclient/sshclient.cockpit.lan'")
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa host-allow-retrieve-keytab --user=admin sshserver.cockpit.lan'")
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:62]: Why does that not have a "servicename/" prefix?

File: test/verify/check-s4u-ssh [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+principal = gssapi.Name('cockpitclient/sshclient.cockpit.lan@COCKPIT.LAN', gssapi.NameType.kerberos_principal)
+credResult = gssapi.creds.rcred_cred_store.acquire_cred_from(store = { 'keytab'.encode(): '/etc/cockpit-client.keytab'.encode() }, name=principal)
+impersonee = gssapi.Name('user@COCKPIT.LAN', gssapi.NameType.kerberos_principal)
+credResults4u2self = gssapi.creds.rcred_s4u.acquire_cred_impersonate_name(credResult.creds, impersonee)
+gssapi.creds.rcred_cred_store.store_cred_into({ 'ccache'.encode(): 'FILE:/etc/cockpit.ccache'.encode(), 'keytab'.encode(): '/etc/cockpit-client.keytab'.encode() }, credResults4u2self.creds, overwrite=True)
+""")
+        client_machine.execute("python /tmp/make-cache.py")
+
+        # enable gssapiauth
+        sshd_machine.execute("sed -i 's/#GSSAPIAuthentication yes/GSSAPIAuthentication yes/' /etc/ssh/sshd_config")
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:89]: Perhaps on Debian/Ubuntu etc. the default file looks differently and doesn't have a commented-out version of this? Also, on latest  Fedora and RHEL, ssh now supports /etc/ssh/ssh_config.d/ , which is rather convenient for cases like this. But I suppose that's not available on older OSes yet :sob:

File: test/verify/check-s4u-ssh (hunk 3 above)
Comment (Gundersanne) [reply test/verify/check-s4u-ssh:66]: nothing actually, the services refresh has the fix, i'll remove this comment ^_^

File: test/verify/check-s4u-ssh (hunk 4 above)
Comment (Gundersanne) [reply test/verify/check-s4u-ssh:70]: yup

File: test/verify/check-s4u-ssh (hunk 5 above)
Comment (Gundersanne) [reply test/verify/check-s4u-ssh:57]: in order to kinit as user using a password you'd need to execute this, but the s4u should work without it so i'm not sure

File: test/verify/check-s4u-ssh (hunk 7 above)
Comment (Gundersanne) [reply test/verify/check-s4u-ssh:62]: it's `host-allow-retrieve-keytab` which allows you to retrieve the keytab for a host.

Comment (martinpitt) [reply test/verify/check-s4u-ssh:62]: D'oh, of course. -10 points from House Augsburg for lack of reading skills!

File: test/verify/check-s4u-ssh (hunk 2 above)
Comment (Gundersanne) [reply test/verify/check-s4u-ssh:64]: I did not know that was a thing, i'll change that :)

File: test/verify/check-s4u-ssh (hunk 7 above)
Comment (Gundersanne) [reply test/verify/check-s4u-ssh:62]: haha it's ok, it's not actually needed as the server doesn't need to kinit at all. So I can just remove it :)

File: test/verify/check-s4u-ssh (hunk 8 above)
Comment (Gundersanne) [reply test/verify/check-s4u-ssh:89]: I think the default there is "#GSSAPIAuthentication no" :D

File: test/verify/check-s4u-ssh [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+import parent
+
+from testlib import *
+
+@skipImage("No realmd available", "fedora-coreos")
+@skipImage("freeipa not currently in testing", "debian-testing")
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:25]: Not in ubuntu-2004 either, see check-realms.

File: test/verify/check-s4u-ssh (hunk 6 above)
Comment (Gundersanne) [reply test/verify/check-s4u-ssh:59]: It means the admin is allowed to retrieve the keytab for that service, so we can do it later in line.

**Comment by martinpitt** [comment]
https://github.com/cockpit-project/bots/pull/1050 landed, retrying debian-stable. This still needs at least one more force-push for ubuntu-2004 though.

File: test/verify/check-s4u-ssh [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        # configure ipa
+        ipa_machine.execute("docker exec freeipa bash -c 'echo foobarfoo | kinit admin@COCKPIT.LAN'")
+        # add user to impersonate
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa user-add --first=user --last=user user'")
+        # add service which will be allowed to delegate creds
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa service-add cockpitclient/sshclient.cockpit.lan@COCKPIT.LAN --ok-as-delegate=true --ok-to-auth-as-delegate=true'")
+        # Allow restrieval of service/host keytabs by admin
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa service-allow-retrieve-keytab --user=admin cockpitclient/sshclient.cockpit.lan'")
+        # set up delegation rule
+        ipa_machine.execute(script="""#!/bin/sh
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:62]: I wasn't even aware of `script=` :grin: 

I actually meant something like
```
script = """ipa service ....
ipa ...
"""
ipa_machine.execute("docker exec freeipa bash -ec '%s'" % script)
```

Like this it also avoids most of the overhead, but at least needs a `set -e`.

**Comment by martinpitt** [comment]
I ran

   test/image-prepare -q ubuntu-stable
    bots/image-customize -i gssproxy -v ubuntu-stable
    TEST_OS=ubuntu-stable test/verify/check-s4u-ssh TestS4USsh.testBasic -stv

and that does *not* fix it, it fails in the same way. Also, the `gssproxy` package is not installed on debian-stable, so it's not that.

File: test/verify/check-s4u-ssh (hunk 10 above)
Comment (Gundersanne) [reply test/verify/check-s4u-ssh:62]: oh hahaha, i'll do that.

And ye I have the other fixes locally.

File: test/verify/check-s4u-ssh [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+import parent
+
+from testlib import *
+
+@skipImage("No realmd available", "fedora-coreos")
+@skipImage("freeipa not currently in testing", "debian-testing", "ubuntu-2004")
+@skipImage("Skip ubuntu-these for now", "ubuntu-stable", "debian-stable")
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:26]: That reads funny.

File: test/verify/check-s4u-ssh [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        # configure ipa
+        ipa_machine.execute("docker exec freeipa bash -c 'echo foobarfoo | kinit admin@COCKPIT.LAN'")
+        # add user to impersonate
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa user-add --first=user --last=user user'")
+        # add service which will be allowed to delegate creds
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa service-add cockpitclient/sshclient.cockpit.lan@COCKPIT.LAN --ok-as-delegate=true --ok-to-auth-as-delegate=true'")
+        # Allow retrieval of service keytab by admin
+        ipa_machine.execute("docker exec freeipa bash -c 'ipa service-allow-retrieve-keytab --user=admin cockpitclient/sshclient.cockpit.lan'")
+        # set up delegation rule
+        script="""set -e
```
Comment (martinpitt) [inline test/verify/check-s4u-ssh:63]: `set -e` is redundant with `bash -e`, but no harm. Was there a reason why half of the script now runs in one go, and the other half in individual commands?

Comment (Gundersanne) [reply test/verify/check-s4u-ssh:63]: well I want to have a comment per thing that I was doing essentially, I can change it if you like

File: test/verify/check-s4u-ssh (hunk 11 above)
Comment (Gundersanne) [reply test/verify/check-s4u-ssh]: indeed it does, already changed it :)

File: test/verify/check-s4u-ssh (hunk 12 above)
Comment (martinpitt) [reply test/verify/check-s4u-ssh:63]: Sure, comments can also go into the `script=` constant. But either way, not important, I was just curious.

**Pull request merged**